### PR TITLE
Special case integer key sizes

### PR DESCRIPTION
**Pull request by Alan-Jowett**
## Description

This pull request includes significant changes to the `libs/runtime/ebpf_hash_table.c` file, primarily focusing on improving the way keys are compared in the hash table. The changes involve the introduction of new comparison methods for different types of keys and a wrapper function to select the best comparison method based on the key size and extract function.

Key comparison improvements:

* [`_ebpf_hash_table_compare_integer_keys`](diffhunk://#diff-6c24a75cf729f18fc4eb4459a9eb23a3561b8192157748b256c40db9abc47c2aL158-R227): A new function that compares keys assuming they are integers. It handles different key sizes and returns a comparison result based on the values of the keys.
* [`_ebpf_hash_table_compare_extracted_keys`](diffhunk://#diff-6c24a75cf729f18fc4eb4459a9eb23a3561b8192157748b256c40db9abc47c2aL158-R227): Another new function that compares keys that require extraction and compares them as byte arrays. This function replaces the original `_ebpf_hash_table_compare` function.
* [`_ebpf_hash_table_compare`](diffhunk://#diff-6c24a75cf729f18fc4eb4459a9eb23a3561b8192157748b256c40db9abc47c2aR246-R270): This function has been repurposed as a wrapper to select the best comparison method for keys based on the key size and extract function. It uses the two new comparison methods mentioned above, or defaults to comparing keys as byte arrays.

## Testing

CI/CD and performance testing:
Baseline:
[01:31:25] :: 2024-07-16T08:31:24-0800,BPF_MAP_TYPE_HASH read,174,175,174,173,174

This change:
[09:35:24] :: 2024-07-17T16:35:23-0800,BPF_MAP_TYPE_HASH read,133,133,133,133,134

## Documentation

No.

## Installation

No.

### CODE REVIEW

File: libs/runtime/ebpf_hash_table.c [hunk 1]
```diff
@@ -189,6 +189,21 @@ _ebpf_hash_table_compare(_In_ const ebpf_hash_table_t* hash_table, _In_ const ui
...
+        // Special case common key sizes.
+        switch (length_a_in_bits) {
+        case 8:
+            return (uint8_t)*data_a - (uint8_t)*data_b;
+        case 16:
+            return (uint16_t) * (uint16_t*)data_a - (uint16_t) * (uint16_t*)data_b;
+        case 32:
+            return (uint32_t) * (uint32_t*)data_a - (uint32_t) * (uint32_t*)data_b;
+        case 64:
+            return (uint32_t)((uint64_t) * (uint64_t*)data_a - (uint64_t) * (uint64_t*)data_b);
```
Comment (saxena-anurag) [inline libs/runtime/ebpf_hash_table.c:203]: since we are truncating to 32 bits, it can incorrectly return 0 if lower 32 bits of the difference are 0?

Comment (Alan-Jowett) [reply libs/runtime/ebpf_hash_table.c:203]: Good point.

File: libs/runtime/ebpf_hash_table.c [hunk 2]
```diff
@@ -155,41 +155,76 @@ _ebpf_murmur3_32(_In_reads_((length_in_bits + 7) / 8) const uint8_t* key, size_t
...
  * @param[in] hash_table Hash table the keys belong to.
  * @param[in] key_a First key.
  * @param[in] key_b Second key.
- * @retval -1 if key_a < key_b
- * @retval 0 if key_a == key_b
- * @retval 1 if key_a > key_b
+ *
+ * @return -1 if key_a < key_b, 0 if key_a == key_b, 1 if key_a > key_b.
+ */
+static __forceinline int
```
Comment (saxena-anurag) [inline libs/runtime/ebpf_hash_table.c:166]:
```suggestion
static __forceinline int32_t
```

**Review by saxena-anurag — APPROVED** [review]

File: libs/runtime/ebpf_hash_table.c (hunk 2 above)
Comment (Alan-Jowett) [reply libs/runtime/ebpf_hash_table.c:166]: Given that memcmp is defined as returning int, so I want to keep it the same so the compiler can tail call to memcmp.

**Review by mikeagun — APPROVED** [review]

**Pull request merged**